internal/extend/kaniko/dockerfile_applier.go

```diff
@@ -17,25 +21,65 @@ const (
 )
 
 var (
-	kanikoCacheDir      = filepath.Join(kanikoDir, "cache", "base")
+	kanikoBaseCacheDir  = filepath.Join(kanikoDir, "cache", "base")
 	kanikoCacheImageRef = filepath.Join(ociPrefix, kanikoDir, "cache", "layers", "cached")
 )
 
-type DockerfileApplier struct {
-	logger log.Logger
+type DockerfileApplier struct{}
+
+func (a *DockerfileApplier) ImageFor(reference string) (v1.Image, error) {
+	digest, err := name.NewDigest(reference)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get digest for reference '%s': %w", reference, err)
+	}
+	baseImage, err := readOCI(ociPrefix + filepath.Join(kanikoDir, "cache", "base", digest.DigestStr()))
+	if err != nil {
+		return nil, fmt.Errorf("getting base image for digest '%s': %w", digest, err)
+	}
+	return baseImage, nil
+}
+
+func readOCI(path string) (v1.Image, error) {
+	if !strings.HasPrefix(path, "oci:") {
+		return nil, fmt.Errorf("expected '%s' to have prefix 'oci:'", path)
+	}
+	layoutPath, err := layout.FromPath(strings.TrimPrefix(path, "oci:"))
+	if err != nil {
+		return nil, fmt.Errorf("getting layout from path: %w", err)
+	}
+	hash, err := v1.NewHash(filepath.Base(path))
+	if err != nil {
+		return nil, fmt.Errorf("getting hash from reference '%s': %w", path, err)
+	}
+	v1Image, err := layoutPath.Image(hash) // FIXME: we may want to implement path.Image(h) in the imgutil 'sparse' package so that trying to access layers on this image errors with a helpful message
+	if err != nil {
+		return nil, fmt.Errorf("getting image from hash '%s': %w", hash.String(), err)
+	}
+	return v1Image, nil
 }
 
-func NewDockerfileApplier(logger log.Logger) *DockerfileApplier {
-	return &DockerfileApplier{
-		logger: logger,
+func (a *DockerfileApplier) Cleanup() error {
```

Comment:
This could cause problems when images are extended in parallel (if the kaniko directory is re-used, which it might be). It may be safer to move this logic to the exporter.